rust/oak/src/tests.rs

```diff
@@ -0,0 +1,134 @@
+//
+// Copyright 2018 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+extern crate oak_tests;
+
+use crate::*;
+
+#[test]
+fn test_status_from_i32() {
+    assert_eq!(Status::InvalidArgs, status_from_i32(2));
+    assert_eq!(Status::Unknown(42), status_from_i32(42));
+}
+
+#[test]
+fn test_result_from_status() {
+    assert_matches!(result_from_status(Status::Ok, 12), Ok(12));
+    assert_matches!(result_from_status(Status::Unknown(42), 12), Err(_));
+}
+
+#[test]
+fn test_write_message() {
+    let mut send_channel = SendChannelHalf::new(123);
+    let data = [0x44, 0x4d, 0x44];
+    assert_matches!(send_channel.write_message(&data), Ok(()));
+    assert_eq!("DMD", oak_tests::last_message());
+}
+
+#[test]
+fn test_write_message_failure() {
+    let mut send_channel = SendChannelHalf::new(123);
+    let data = [0x44, 0x4d, 0x44];
+    oak_tests::set_status(Some(99));
+    assert_matches!(send_channel.write_message(&data), Err(_));
+}
+
+#[test]
+fn test_write() {
+    writeln!(logging_channel(), "ABC").unwrap();
+    assert_eq!("ABC\n", oak_tests::last_message());
+    assert_matches!(logging_channel().flush(), Ok(()));
+}
+
+#[test]
+fn test_read_message() {
+    let mut send = SendChannelHalf::new(123);
+    let data = [0x44, 0x4d, 0x44];
+    assert_matches!(send.write_message(&data), Ok(()));
+
+    let mut rcv = ReceiveChannelHalf::new(123);
+    let mut buf = Vec::with_capacity(1);
+    assert_matches!(rcv.read_message(&mut buf), Ok(3));
+    assert_eq!(data.to_vec(), buf);
+
+    let mut big_buf = Vec::with_capacity(100);
+    assert_matches!(rcv.read_message(&mut big_buf), Ok(3));
+    assert_eq!(data.to_vec(), big_buf);
+}
+
+#[test]
+fn test_read_message_failure() {
+    let mut send = SendChannelHalf::new(123);
+    let data = [0x44, 0x4d, 0x44];
+    oak_tests::set_status(Some(99));
```

Comment:
Yeah, the global mutable state is definitely a testing-only hack.  I can certainly encapsulate the test methods into a `MockChannel` but I can't immediately see how to avoid the global mutable aspect &ndash; the calls to `channel_read(1, ...)` in the code under test have no context to get to something non-global.

(BTW, in the future I'm thinking that the global mutable state would evolve to being a `HashMap<i32, MockChannel>` not just a `MockChannel` to cope with multi-channel testing.)